src/google/protobuf/descriptor.cc

```diff
@@ -5894,6 +5894,21 @@ void DescriptorBuilder::ValidateProto3Message(
       name_to_field[lowercase_name] = message->field(i);
     }
   }
+
+  HASH_MAP<string, const FieldDescriptor*> reserved_json_name_map;
```

Comment:
This problem affects proto2 as well, so I'm not sure the `ValidateProto3Message` function is the right place to add this.